### PR TITLE
One-line app registration and model deployment

### PR DESCRIPTION
**Pull request by nishadsingh1**
Addresses the third checkbox in #218

### CODE REVIEW

File: clipper_admin/clipper_manager.py [hunk 1]
```diff
@@ -345,6 +345,12 @@ def register_application(self, name, model, input_type, default_output,
             the objective not be set aggressively low unless absolutely necessary.
             40000 (40ms) is a good starting value, but the optimal latency objective
             will vary depending on the application.
+
+        Returns
+        -------
+        Requests object
+            The response object for the app registration request
```
Comment (dcrankshaw) [inline clipper_admin/clipper_manager.py]: What is a response object?

File: clipper_admin/clipper_manager.py [hunk 2]
```diff
@@ -733,6 +740,80 @@ def centered_predict(inputs):
 
         return deploy_result
 
+    def _register_app_and_check_success(self, app_name, model_name, input_type,
+                                        default_output, slo_micros):
+        response = self.register_application(app_name, model_name, input_type,
+                                             default_output, slo_micros)
+        if response.status_code != 200:
```
Comment (dcrankshaw) [inline clipper_admin/clipper_manager.py]: Don't check for 200, use `if r.status_code == requests.codes.ok:`

File: clipper_admin/clipper_manager.py [hunk 3]
```diff
@@ -733,6 +740,80 @@ def centered_predict(inputs):
...
+            print(
+                "Application registration unsuccessful. Will not deploy model.")
+            return False
+        print("Application registration sucessful! Deploying model.")
+        return True
+
+    def register_app_and_deploy_model(self,
```
Comment (dcrankshaw) [inline clipper_admin/clipper_manager.py]: These are really really long method names.

**Review by dcrankshaw — CHANGES_REQUESTED** [review]
So now that I'm looking at the code, these method signatures still look really complicated. These are basically convenience methods to make things simpler. If the user wants more control, she can call `register_application` and deploy_*` separately.

To simplify the method signatures:
1. Make the model name and app name the same and just have a single name parameter
2. Remove the method for `register_app_and_deploy_model`, just have it for python functions and pyspark models
3. Make the model version arg have a default argument of 1 (`model_version=1`)
4. Make slo_micros default to 100000
5. Make `input_type` the last non-default arg (so the order is name, predict_function, [pyspark model and spark context if deploy_pyspark method], input_type, default args)

**Comment by nishadsingh1** [comment]
Makes sense -- do we want to give `default_output` a default as well? Maybe "-1.0" or something?

**Comment by nishadsingh1** [comment]
Didn't make the change to have a default `default_output` -- happy to add it in. The new function names `activate_predict_function` and `activate_pyspark_model` still don't seem quite right...let me know if better names come to mind.

File: clipper_admin/clipper_manager.py [hunk 4]
```diff
@@ -44,6 +44,8 @@
 CLIPPER_DOCKER_LABEL = "ai.clipper.container.label"
 CLIPPER_MODEL_CONTAINER_LABEL = "ai.clipper.model_container.model_version"
 
+DEFAULT_MODEL_VERSION = 1
+DEFAULT_SLO_MICROS = 10000
```
Comment (dcrankshaw) [inline clipper_admin/clipper_manager.py]: This is 10 ms, make it 100 ms (100000)

File: clipper_admin/clipper_manager.py [hunk 5]
```diff
@@ -345,6 +347,12 @@ def register_application(self, name, model, input_type, default_output,
             the objective not be set aggressively low unless absolutely necessary.
             40000 (40ms) is a good starting value, but the optimal latency objective
             will vary depending on the application.
+
+        Returns
+        -------
+        bool
+            Returns true iff the app reigstration request was successful
```
Comment (dcrankshaw) [inline clipper_admin/clipper_manager.py]: typo in "registration"

File: clipper_admin/clipper_manager.py [hunk 6]
```diff
@@ -735,6 +744,105 @@ def centered_predict(inputs):
...
+    def _register_app_and_check_success(self, name, input_type, default_output,
+                                        slo_micros):
+        if self.register_application(name, name, input_type, default_output,
+                                     slo_micros):
+            print("Application registration sucessful! Deploying model.")
+            return True
+        print("Application registration unsuccessful. Will not deploy model.")
+        return False
+
+    def activate_predict_function(self,
```
Comment (dcrankshaw) [inline clipper_admin/clipper_manager.py]: Activate isn't really the right verb. I can't think of a good shorter name that doesn't conflict with our existing API, so maybe just use the long names from before.

File: clipper_admin/clipper_manager.py [hunk 7]
```diff
@@ -735,6 +744,105 @@ def centered_predict(inputs):
...
+                                  name,
+                                  predict_function,
+                                  default_output,
```
Comment (dcrankshaw) [inline clipper_admin/clipper_manager.py]: Make the default output "None"

File: clipper_admin/clipper_manager.py [hunk 8]
```diff
@@ -735,6 +744,105 @@ def centered_predict(inputs):
...
+        return self.deploy_predict_function(name, model_version,
+                                            predict_function, input_type,
+                                            labels, num_containers)
+
+    def activate_pyspark_model(self,
+                               name,
+                               predict_function,
+                               pyspark_model,
+                               sc,
+                               default_output,
```
Comment (dcrankshaw) [inline clipper_admin/clipper_manager.py]: Make the default output "None"

File: clipper_admin/clipper_manager.py [hunk 9]
```diff
@@ -735,6 +744,105 @@ def centered_predict(inputs):
...
+                                  slo_micros=DEFAULT_SLO_MICROS,
+                                  labels=DEFAULT_LABEL,
+                                  num_containers=1):
+        """Registers an app and deploys provided predict function as a model.
+
+        Parameters
+        ----------
+        name : str
+            The to be assigned to the registered app and deployed model.
+        predict_function : function
```
Comment (dcrankshaw) [inline clipper_admin/clipper_manager.py:773]: Don't say "refer to documentation in ..." That is really confusing and leads to hard to read docs. Just copy the text.

As we refactor this documentation, we'll create some module-level documentation about deploying models and the meaning of these fields.

File: clipper_admin/clipper_manager.py [hunk 10]
```diff
@@ -735,6 +744,105 @@ def centered_predict(inputs):
...
+                               slo_micros=100000,
+                               labels=DEFAULT_LABEL,
+                               num_containers=1):
+        """Registers an app and deploys provided spark model.
+
+        Parameters
+        ----------
+        name : str
+            The to be assigned to the registered app and deployed model.
+        predict_function : function
```
Comment (dcrankshaw) [inline clipper_admin/clipper_manager.py:821]: Fix docs (remove "refer to...")

**Review by dcrankshaw — CHANGES_REQUESTED** [review]
Mostly there. Just some naming and doc cleanup.

**Comment by dcrankshaw** [comment]
LGTM. Merge with develop (looks like you'll need to do that locally) and I'll do a last pass and get this merged.

**Review by dcrankshaw — APPROVED** [review]

**Pull request merged**